mm2src/mm2_net/src/wasm/body_stream.rs

```diff
@@ -0,0 +1,411 @@
+    incomplete_data: BytesMut,
+    data: Option<BytesMut>,
+    trailer: Option<HeaderMap>,
+    state: ReadState,
+    finished_stream: bool,
+}
+
+impl ResponseBody {
+    /// Creates a new `ResponseBody` based on a ReadableStream and content type.
+    pub(crate) async fn new(body_stream: ReadableStream, content_type: &str) -> Result<Self, PostGrpcWebErr> {
+        let body_stream: ReadableStreamDefaultReader = body_stream
+            .get_reader()
+            .dyn_into()
+            .map_err(|err| PostGrpcWebErr::BadResponse(format!("{err:?}")))?;
+
+        Ok(Self {
+            body_stream: BodyStream::new(body_stream).await?,
+            buf: EncodedBytes::new(content_type)?,
+            incomplete_data: BytesMut::new(),
+            data: None,
+            trailer: None,
+            state: ReadState::CompressionFlag,
+            finished_stream: false,
+        })
+    }
+
+    fn read_stream(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), PostGrpcWebErr>> {
+        if self.finished_stream {
+            return Poll::Ready(Ok(()));
+        }
+
+        let this = self.project();
+
+        match ready!(this.body_stream.poll_data(cx)) {
+            Some(Ok(data)) => {
+                if let Err(e) = this.buf.append(data) {
+                    return Poll::Ready(Err(e));
+                }
+
+                Poll::Ready(Ok(()))
+            },
+            Some(Err(e)) => Poll::Ready(Err(e)),
+            None => {
+                *this.finished_stream = true;
+                Poll::Ready(Ok(()))
+            },
+        }
+    }
+
+    fn step(self: Pin<&mut Self>) -> Result<(), PostGrpcWebErr> {
+        let this = self.project();
+
+        loop {
+            match this.state {
+                ReadState::CompressionFlag => {
+                    if this.buf.is_empty() {
+                        // Can't read compression flag right now
+                        return Ok(());
+                    } else {
+                        let compression_flag = this.buf.take(1);
+
+                        if compression_flag[0] & TRAILER_BIT == 0 {
+                            this.incomplete_data.unsplit(compression_flag);
+                            *this.state = ReadState::DataLength;
+                        } else {
+                            *this.state = ReadState::TrailerLength;
+                        }
+                    }
+                },
+                ReadState::DataLength => {
+                    if this.buf.len() < 4 {
+                        // Can't read data length right now
+                        return Ok(());
+                    } else {
+                        let data_length_bytes = this.buf.take(4);
+                        let data_length = BigEndian::read_u32(data_length_bytes.as_ref());
+
+                        this.incomplete_data.extend_from_slice(&data_length_bytes);
+                        *this.state = ReadState::Data(data_length);
+                    }
```

Comment:
Same here and in similar if else statements below

Reply:
done